tweaks, basic DWARF-based info

monoidic opened the pull request:
This PR sets up the groundwork for getting DWARF-based info from ELF files, and adds fetching the built-in goroot and build version using it as examples, which currently otherwise require hacks like disassembly of code. The addresses and values of various other values (e.g `runtime.firstmoduledata` for the moduledata, or various type definitions) could be located as well.
This, of course, has the drawback of only working on ELF and requiring DWARF to not be stripped, though it should be portable across all Go platforms which use ELF, both by OS and CPU architecture. And DWARF is, by default, included.

This PR also includes some housekeeping and tweaks:
* bumping dependencies
* simplifying `typeOffset()`
* renaming `getSectionDataFromOffset` to `getSectionDataFromAddress` for clarity (the "offset" passed as an argument is a memory address, not a file offset)
* replace uses of the deprecated `io/ioutil` module with equivalent code
* replace some `f.Seek(x, 0)` calls with `f.Seek(x, io.SeekStart)` to remove the magic value
* run `go generate` to update `gopkg_gen.go` and `stdpkg_gen.go`
* fix some typoes
* do not silently ignore some error values
* update tests to handle go.12.0 correctly and add said version to `testdata/build.go` (todo: PR to [/goretk/testdata](/goretk/testdata))

monoidic commented:
And autogenerating the code for the `moduledata{version}{machine_word_size}` structs + converting them to `moduledata` without using any (runtime) reflection.
I also took the liberty of trimming the definitions down to the last actually used field. Go 1.21 changed the definition of the `moduledata` struct, but until `inittasks` itself (which could be interesting) or any of the fields succeeding it are used, it won't make any difference here.

TcM1911 commented:
Thank you for contributing. I will make time to review this by the end of this week.

TcM1911 commented on elf.go:
```diff
@@ -167,3 +170,163 @@ func (e *elfFile) getBuildID() (string, error) {
 }
+
+const (
+	DW_AT_go_kind dwarf.Attr = 0x2900 + iota
```
Let's not export these constants. Also, can you use camel case please. Adding documentation from where this was obtain could be useful in case these values ever changes in the future.

monoidic replied:
The `DW_AT_go_*` constants are defined in `cmd/internal/dwarf/dwarf.go` in the Go source code. However, these constants didn't end up being actually used in any of the code in this PR, so I've removed them for now, and added info on the remaining two.

TcM1911 commented on elf.go:
```diff
@@ -167,3 +170,163 @@ func (e *elfFile) getBuildID() (string, error) {
+	dwFound
+)
+
+func getValFromDwarf[T any](e *elfFile, check func(e *elfFile, entry *dwarfEntryPlus) (T, dwarfwalkStatus)) (T, bool) {
```
I don't see a need of making this function generic. Only `getDwarfStringCheck` exist.

TcM1911 commented on elf.go:
```diff
@@ -167,3 +170,163 @@ func (e *elfFile) getBuildID() (string, error) {
+			case dwStop:
+				break getValOuter
+			case dwFound:
+				fmt.Printf("returning %+v\n", ret)
```
Remove

TcM1911 commented on elf.go:
```diff
@@ -167,3 +170,163 @@ func (e *elfFile) getBuildID() (string, error) {
+func commonStringCheck(e *elfFile, entry *dwarf.Entry) (string, dwarfwalkStatus) {
+	locationField := entry.AttrField(dwarf.AttrLocation)
+	if locationField == nil {
+		// ???
```
Remove or provide better comment.

TcM1911 commented on file_test.go:
```diff
@@ -42,13 +42,13 @@ func TestIssue11NoNoteSectionELF(t *testing.T) {
 		panic("No go tool chain found: " + err.Error())
 	}
-	tmpdir, err := ioutil.TempDir("", "TestGORE-Issue11")
+	tmpdir, err := os.MkdirTemp("", "TestGORE-Issue11")
```
I think the better change is to use the temp dir function provided by `t` since it will be removed automatically once the tests are done.

TcM1911 commented on gen_moduledata.go:
Let's call this file `moduledata.go`.

monoidic replied:
`moduledata.go` already exists.

TcM1911 commented on gen_moduledata.go:
```diff
@@ -0,0 +1,331 @@
+//go:build ignore
+// +build ignore
+
+// This program generates stdpkgs_gen.go and goversion_gen.go. It can be invoked by running
```
Wrong documentation...

TcM1911 commented on goroot.go:
```diff
@@ -344,6 +344,13 @@ func findGoRootPath(f *GoFile) (string, error) {
 		return "", err
 	}
 
+	if e, ok := f.fh.(*elfFile); ok {
```
Should we try this approach first?

monoidic replied:
Seems reasonable.

TcM1911 commented on goversion.go:
```diff
@@ -81,6 +81,15 @@ func findGoCompilerVersion(f *GoFile) (*GoVersion, error) {
 		return v, nil
 	}
 
+	if e, ok := f.fh.(*elfFile); ok {
```
Should we try this approach first?

TcM1911 commented on slow_test.go:
```diff
@@ -448,13 +447,13 @@ func buildTestResource(body, goos, arch string, pie bool) (string, string) {
 	}
 
-	tmpdir, err := ioutil.TempDir("", "TestGORE")
+	tmpdir, err := os.MkdirTemp("", "TestGORE")
```
Same as previous comment.

Zxilly commented:
Any further work on this patch? I can help to merge this.

TcM1911 commented:
The conflicts must be fixed. It would probably be best also to break out the moduledata and the DWARF extraction into separate PRs.

monoidic commented:
Oh, I'd forgotten about this PR.

> The conflicts must be fixed. It would probably be best also to break out the moduledata and the DWARF extraction into separate PRs.

Very well, I will wait for #77 to be accepted and then rework this PR.

TcM1911 commented:
#77 has been merged.

Zxilly commented on testdata/gold:
This part may not necessary.

monoidic replied:
Whoops.

Zxilly commented on moduledata.go:
```diff
@@ -24,10 +24,11 @@ import (
 	"strconv"
 	"strings"
+
+	"golang.org/x/mod/semver"
```
Could you keep the legacy style here? This can prevent the possible conflict with #85

monoidic replied:
`gofumpt` did it, I suppose I can keep the original form.

Zxilly commented:
Maybe we can add some sample with debug info to gold.

Zxilly commented:
I noticed `macho.File` also has a https://pkg.go.dev/debug/macho#File.DWARF method, can this logic also applied to macho?

Zxilly commented:
Even `debug/pe` also has a https://pkg.go.dev/debug/pe#File.DWARF. And all of them use the `debug/dwarf`.
This made me curious, as I always thought pe files used a separate PDB debug info file

monoidic commented:
Huh, you're right, @Zxilly . Somewhat surprised to see that PE is on that list.

I gave the `fileHandler` interface a `getDwarf` method to get the DWARF data and just made the DWARF data extraction code use `fileHandler`, so it works across all binary formats.

Zxilly reviewed (approved):
lgtm

Zxilly commented:
hope this PR get merged soon as it may conflicted with #86, blocking the further development

monoidic committed:
simplyify typeOffset

monoidic committed:
tweaks

- rename `getSectionDataFromOffset` to `getSectionDataFromAddress` for clarity
- use `io.SeekStart` instead of magic value 0
- fix some typoes

monoidic committed:
use DWARF info, if available, to get goroot and goversion

monoidic commented:
Rebased to fix conflicts.

TcM1911 reviewed (changes requested):
There is a lot of places where it results in a panic. Since this is a library, I think we should try to not bring down the main application if we can. Please switch to errors instead.

It could be good to add some tests too. Either by using "golden" binaries or building some on the fly like it's done in https://github.com/goretk/gore/blob/develop/slow_test.go.

monoidic committed:
remove left-over panics

monoidic committed:
tests for DWARF data

monoidic commented:
Those panics were definitely not intended to be there, whoops.
I thought I'd have to have a chicken-or-egg issue with pushing unstripped test binaries to the testdata. But `slow_test` works.
I have to use a different source file if I want `getGoRootFromDwarf` to succeed as `runtime.defaultGOROOT` is not included unless needed.

Also, the `tmpDirs` slice in `TestMain` in `slow_test` was not previously collecting all temporary directories and a bunch of build files were getting left behind. I fixed that.

TcM1911 commented on dwarf.go:
```diff
@@ -0,0 +1,161 @@
+package gore
```
Copyright and license.

TcM1911 reviewed (changes requested):
Minor fix needed and this is ready to be merged.

Zxilly commented:
The CI error was caused by go 1.22, which moduledata currently not exist in the main branch. After https://github.com/goretk/gore/pull/88 merged it gets resolved.

TcM1911 commented:
#88 has been merged. Please try again.

monoidic committed:
add license header to dwarf.go

Zxilly commented:
Seems works fine now.
